static/ast-view.js

```diff
@@ -0,0 +1,135 @@
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+define(function (require) {
+    "use strict";
+
+    var FontScale = require('fontscale');
+    var monaco = require('monaco');
+    var options = require('options');
+
+    require('asm-mode');
+    require('selectize');
+
+    function Ast(hub, container, state) {
+        this.container = container;
+        this.eventHub = hub.createEventHub();
+        this.domRoot = container.getElement();
+        this.domRoot.html($('#ast').html());
+        this.compilers = {};
+        this._currentDecorations = [];
+        this.astEditor = monaco.editor.create(this.domRoot.find(".monaco-placeholder")[0], {
+            value: "",
+            scrollBeyondLastLine: false,
+            language: 'cpp', //we only support cpp for now
+            readOnly: true,
+            glyphMargin: true,
+            quickSuggestions: false,
+            fixedOverflowWidgets: true
+        });
+
+        this._compilerid = state.id;
+        this._compilerName = state.compilerName;
+        this._editorid = state.editorid;
+        this.fontScale = new FontScale(this.domRoot, state, this.astEditor);
+        this.fontScale.on('change', _.bind(this.updateState, this));
+
+        this.eventHub.on('compileResult', this.onCompileResult, this);
+        this.eventHub.on('compiler', this.onCompiler, this);
+        this.eventHub.on('compilerClose', this.onCompilerClose, this);
+        this.eventHub.on('editorChange', this.onEditorChange, this);
+        this.eventHub.on('themeChange', this.onThemeChange, this);
+        this.eventHub.emit('requestTheme');
+        this.eventHub.emit('astViewOpened', this._compilerid);
```

Comment:
Sorry for all this awkward tricky stuff with emitting etc. The message passing library in GoldenLayout pushes us in this direction. In theory one day we can use this to "pop out" windows...but I haven't ever got that working...